File: tests/test_franchise_walk.py
#!/usr/bin/env python3
"""Tests du walk franchise (contrat commun front JS / pipeline Python).
Run: python tests/test_franchise_walk.py
"""
import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from pipelines.anime_tracker_sync import build_franchise, chain_ids, missing_ids

FIXTURES = json.loads((Path(__file__).parent / "fixtures" / "franchise_graphs.json").read_text(encoding="utf-8"))


def media_map(case):
    return {m["id"]: m for m in case["media"]}


def main():
    failures = 0
    for case in FIXTURES["cases"]:
        got = build_franchise(media_map(case), case["anchor"])
        exp = case["expected"]
        if got != exp:
            failures += 1
            print(f"FAIL {case['name']}\n  expected: {exp}\n  got:      {got}")
        else:
            print(f"ok   {case['name']}")
    for mc in FIXTURES["missing_cases"]:
        all_media = None
        for case in FIXTURES["cases"]:
            if mc["anchor"] in media_map(case):
                all_media = media_map(case)
                break
        known = {k: v for k, v in all_media.items() if k in set(mc["known"])}
        got = sorted(missing_ids(known, mc["anchor"]))
        if got != sorted(mc["expected_missing"]):
            failures += 1
            print(f"FAIL {mc['name']}: expected {mc['expected_missing']}, got {got}")
        else:
            print(f"ok   {mc['name']}")
    if failures:
        print(f"\n{failures} failure(s)")
        sys.exit(1)
    print("\nAll walk tests passed.")


if __name__ == "__main__":
    main()

File: tests/test_franchise_walk.mjs
// Miroir JS des tests Python — mêmes fixtures, mêmes attentes.
// Run: node tests/test_franchise_walk.mjs
import { readFileSync } from "node:fs";
import { createRequire } from "node:module";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

const here = dirname(fileURLToPath(import.meta.url));
const require = createRequire(import.meta.url);
const { buildFranchise, missingIds } = require(join(here, "..", "cockpit", "lib", "anilist.js"));
const FIXTURES = JSON.parse(readFileSync(join(here, "fixtures", "franchise_graphs.json"), "utf-8"));

const mediaMap = (c) => Object.fromEntries(c.media.map((m) => [m.id, m]));
const deepEq = (a, b) => JSON.stringify(a) === JSON.stringify(b);

let failures = 0;
for (const c of FIXTURES.cases) {
  const got = buildFranchise(mediaMap(c), c.anchor);
  if (!deepEq(got, c.expected)) {
    failures++;
    console.log(`FAIL ${c.name}\n  expected: ${JSON.stringify(c.expected)}\n  got:      ${JSON.stringify(got)}`);
  } else console.log(`ok   ${c.name}`);
}
for (const mc of FIXTURES.missing_cases) {
  const full = mediaMap(FIXTURES.cases.find((c) => mediaMap(c)[mc.anchor]));
  const known = Object.fromEntries(Object.entries(full).filter(([k]) => mc.known.includes(Number(k))));
  const got = [...missingIds(known, mc.anchor)].sort((a, b) => a - b);
  if (!deepEq(got, [...mc.expected_missing].sort((a, b) => a - b))) {
    failures++;
    console.log(`FAIL ${mc.name}: expected ${mc.expected_missing}, got ${got}`);
  } else console.log(`ok   ${mc.name}`);
}
if (failures) { console.log(`\n${failures} failure(s)`); process.exit(1); }
console.log("\nAll JS walk tests passed.");
